Doodads Refactor

SoulMuncher opened the pull request:
Closes #97 
- Urn shows explosion radius when selected
- Going to revisit this code in broader context, and look for a solution to #98 

Closes #71 
- Urn corpses are now correctly removed, fixing tint and blood trail issues

QA Single/Multi/Loading

SoulMuncher commented:
I wonder if there's a better way to do Unit.SubType

Unit.SubType is essentially used to define a bunch of specific behaviors for similar types of units. Doodads might have a lot of different behaviors.

Maybe instead of specifying broader types, we specify the behaviors themselves, such as unit targeting style, whether or not it leaves a corpse, what targeting category it's in and so on.

jdoleary commented:
> I wonder if there's a better way to do Unit.SubType
> 
> Unit.SubType is essentially used to define a bunch of specific behaviors for similar types of units. Different doodads might have a lot of different behaviors.
> 
> Maybe instead of specifying broader types, we specify the behaviors themselves, such as unit targeting style, whether or not it leaves a corpse, what targeting category it's in and so on.

That is a very interesting idea.  I think i'd have to see the overlaps laid out to feel like a refactor like that would be worth it tho.  For now it's basically attack style with a few exceptions.

SoulMuncher commented:
@jdoleary 
I made a bunch of other changes stemming off of this branch, and now that I've made them I realize they probably should've been a separate branch/pr since it's a pretty big one and addresses a lot of unrelated issues. I'll wait until this is merged and open up a new PR branched off of the new "master". It's a big one 👀

SoulMuncher commented:
These last few commits greatly improve many game visuals.

- Added a new unit event: onDrawSelected. This lets you define specific drawing behaviors for different units, such as drawing a line for LOS units or an effect radius for doodads. What's extra cool is this also lets you attach modifiers with their own drawing properties (such as a bloat explosion radius) to appear when the unit is selected.
- Units now show a bloat explosion radius when selected
- Fixed an issue that cause most floating text, unit markers, and other "world UI" to leave the screen border at various zoom levels
- Fixed an issue that caused some prediction visuals and their label text to flicker when moving the mouse
- Clarity of floating text has been increased significantly (still room to change the colors and dropdown shadow if requested)
- Centralized a lot of draw functions, improving visual consistency and code maintainability

Closes #98 

IMPORTANT: The new unit event causes old saves to not load properly. This probably needs an exception handler or something. Let me know what I need to do for that, if anything 👍 

QA'd Single/Multi/Loading

SoulMuncher commented:
Closes #96
- Unit Hp/Mp bars now show "healing"
- Also have transparent backgrounds

jdoleary commented on src/entity/Unit.ts:
```diff
@@ -1444,6 +1450,91 @@ export function findLOSLocation(unit: IUnit, target: Vec2, underworld: Underworl
+      if (drawEvent) {
+        const fn = Events.onDrawSelectedSource[drawEvent];
+        if (fn) {
+          await fn(unit, prediction, underworld);
```
Does this event function need to be async?

SoulMuncher replied:
Oh!... Probably not.

jdoleary commented on src/graphics/PlanningView.ts:
```diff
@@ -43,31 +43,33 @@ export function initPlanningView() {
       containerUI.addChild(labelText);
     }
     if (mouseLabelText) {
+      mouseLabelText.style.fontSize = 100;
```
The floating text looks great! All it took was increasing the font size and shrinking the scale??

SoulMuncher replied:
Yeah pretty much! The label text previously had a font size of ~26 I think

jdoleary commented on src/entity/Unit.ts:
```diff
@@ -1444,6 +1450,91 @@ export function findLOSLocation(unit: IUnit, target: Vec2, underworld: Underworl
+// handles drawing attack range, bloat radius, and similar graphics each frame while a unit is selected
+export async function drawSelectedGraphics(unit: IUnit, prediction: boolean = false, underworld: Underworld): Promise<void> {
+
+  if (!prediction) {
```
Let's have this exclude globalThis.headless too.  It shouldn't cause any errors but there's no reason for the headless server to have to take any of the branches in this function since it's visual only.

jdoleary commented on src/graphics/PlanningView.ts:
```diff
@@ -483,88 +335,168 @@ export function drawHealthBarAboveHead(unitIndex: number, underworld: Underworld
+        const healthBarHealColor = u.faction == Faction.ALLY ? colors.healthAllyBrightGreen : colors.healthBrightRed;
+
+        const healthBarMax = u.healthMax || 1;
+        //background
```
The healthbar background seems to apply to the player but not to golems

jdoleary replied:
Edit: I was wrong, it does show on NPCs but for the player the background color is the samee as the "Prediction Loss Health" color whereas for the NPCs it's not.  I think it'd be best if there were a different color for each.  For allies, real health is the bright green, prediction loss health is the dark green and then missing health could be the very dark grey (non colored). And for enemies real health is red, prediction loss health is dark red and missing health (background) is dark grey?

SoulMuncher replied:
The background is dark grey currentlt. I think what you're seeing is the transparency, making it mix with color behind it.

I thought it looked a little strange too. Now I'm more confident it making it somewhat (or completely) opaque.

Also, you might've noticed non-mana units do not have a mana bar background. And if a non-mana unit does get mana, it will show as a full bar. I thought that was especially nice for send mana 🙂

jdoleary commented on src/graphics/PlanningView.ts:
```diff
@@ -483,88 +335,168 @@ export function drawHealthBarAboveHead(unitIndex: number, underworld: Underworld
+  const markerHeightHalf = 16 * markerScale;
+  const markerMarginAboveHealthBar = 10;
+
+  // TODO - move marker higher if the unit is a miniboss
```
You should be able to use unit.image.scale.y to determine this since minibosses and stacked summoned units are just scaled up.

SoulMuncher replied:
Another TODO I forgot to remove haha. Had to jump around a lot and do intermittent testing with these changes. Will clean it up shortly.

jdoleary commented on src/graphics/PlanningView.ts:
```diff
@@ -694,6 +626,31 @@ export async function runPredictions(underworld: Underworld) {
+  }
+  if (globalThis.radiusGraphics) {
+    for (let { target, color, radius, text } of predictionCirclesFill) {
+      //const colorOverride = currentlyWarningOutOfRange ? colors.outOfRangeGrey : color;
```
Prediction bloat is red instead of grey when enemy is out of range.  It should be grey to clearly denote to the user that this  _would_ be their bloat explosion radius but they're out of range and can't be cast on

jdoleary replied:
I'm not sure if this is the area where that went wrong though, just noticed that it's commented out and might be an issue here too

jdoleary commented on src/graphics/ui/eventListeners.ts:
```diff
@@ -851,6 +850,9 @@ export function clickHandler(overworld: Overworld, e: MouseEvent) {
   } else {
+
+    const mouseTarget = underworld.getMousePos();
+    drawCircleUnderTarget(mouseTarget, underworld, 1.0, globalThis.planningViewGraphics);
```
What is this for? Click handler just triggers once on click and then is done.  Persistent circles on screen are usually drawn inside of recurring functions such as gameLoop or something that is called every frame

SoulMuncher replied:
Um... I'm not really sure. In some situations where units and pickups were nearby, there was a discrepancy between what highlighted when you hovered, and what you actually selected when you clicked. I think I was testing something here and forgot to remove it.

jdoleary reviewed (commented):
Looking really good! A few comments, questions and requested changes.

Also,
<img width="962" alt="image" src="https://github.com/jdoleary/Spellmasons/assets/6413317/017f1167-8fd6-49c6-9f85-60503ccc4cbc">
Attack range is labelled but bloat isn't.

jdoleary commented:
> IMPORTANT: The new unit event causes old saves to not load properly. This probably needs an exception handler or something. Let me know what I need to do for that, if anything 👍

To ensure backwards compatibility, all we have to do is supply a default for if the new object or array is undefined: 
for example,
`onDrawSelectedEvents: [...(onDrawSelectedEvents || [])],` This will spread the empty array if onDrawSelectedEvents is falsey.
I think there's 2 different places where this needs to be fixed.  The other is ...rest.onDrawSelectedEvents.

SoulMuncher commented:
> Attack range is labelled but bloat isn't.

There is only ever one "label text" at a time from the planning view, which includes the selectedunit and prediction visuals. I think this was something you did intentionally to prevent the screen getting too messy when you do something like [connect-circle-connect-circle...]

What's the expected behavior here? I can't decide if we should show everything or do something else

SoulMuncher commented on src/Underworld.ts:
```diff
@@ -512,6 +503,7 @@ export default class Underworld {
         startPos.y += unitImageYOffset;
         const done = this.runForceMove(forceMoveInst, false);
         const endPos = { x: forceMoveInst.pushedObject.x, y: forceMoveInst.pushedObject.y + unitImageYOffset };
+        // TODO - Fix urn blood smear
```
This was fixed, I need to removed the TODO

jdoleary commented:
> > Attack range is labelled but bloat isn't.
> 
> There is only ever one "label text" at a time from the planning view, which includes the selectedunit and prediction visuals. I think this was something you did intentionally to prevent the screen getting too messy when you do something like [connect-circle-connect-circle...]
> 
> What's the expected behavior here? I can't decide if we should show everything or do something else

Yeah I did do this intentionally.  Most units are going to have an attack radius, so won't that mean that selecting them will always no show the "bloat explosion radius" label?  I think if the label isn't visible the circle shouldn't show because it's not obvious what it is then.

jdoleary commented on src/entity/Unit.ts:
```diff
@@ -1452,25 +1452,23 @@ export function findLOSLocation(unit: IUnit, target: Vec2, underworld: Underworl
-    console.error("selectedUnitGraphics is undefined")
-    return;
-  }
+  // TODO - Ideally the logic below would be defined in each Unit's ts file individually
+  // Instead of using if/else and unit subtypes
```
Why? If the logic is always the same for each given subtype wouldn't that just result in a lot of repetition for units that share the same subtype?

SoulMuncher replied:
This would be for the same reason we put .action or .getUnitAttackTargets on each unit instead of in a large if/else. 

There are already many different ways a unit can draw information, and I imagine there will only be more as new units are added, especially when considering different kinds of doodads, aoe enemies, spellcasters, and so on. Many of the current drawings could be separated and improved too, such as the glops and priests. It would be nice to show more than just the attack range circle for most enemies, and I think that defining it in each enemy's script would be a cleaner way of doing this than to have the unit specific logic be in the broader unit class (and onDrawSelected is currently the only place this happens, afaik)

It also separates drawing behavior from unit.subtypes, making the subtypes more of an identification/targeting group thing, which I think is another step in the right direction.

jdoleary replied:
Yeah that sounds good to me

SoulMuncher commented:
> Most units are going to have an attack radius, so won't that mean that selecting them will always no show the "bloat explosion radius" label? I think if the label isn't visible the circle shouldn't show because it's not obvious what it is then.

You're right that the bloat radius text is always hidden because the planning view is limited to one label currently. I don't think removing the circle is a good idea though. Like you said, the issue is that it's not obvious what the circle is, not that the circle exists in the first place. In a broader context, the whole point of the new onDraw system is to let players see more relevant information about the selected unit and to be able to make draw calls with modifiers. The problem/solution is finding a clean, clear, and concise way to convey that information and make those draw calls.

Even without any changes, I think that an unlabeled bloat circle is better than a non-existent one. If the player knows they put bloat on a unit, they'll figure out what the circle is, and if they don't know they put bloat on a unit, it's even more important that the circle is there to warn them imo. Especially when predictions don't consider ally moves.

Do you have any ideas for handling the label text? I could make a new issue out of it if it's helpful 👍

SoulMuncher commented:
> To ensure backwards compatibility

I might need some help with this. I am still running into issues on selection/bloat application for a v1.24 save. Think you can take a look at it?

jdoleary commented:
The new hp bar background looks great, and the trick with giving melee units a mana bar if they receive mana is brilliant.  Works exactly as i'd expect it to.
I'm going to merge this and I'll create a new issue to revisit the bloat label problem, but I want to get this one it since it was a ton of work!

jdoleary reviewed (approved):
Great work!
